src/VisualStudio/Core/Def/Implementation/DesignerAttribute/VisualStudioDesignerAttributeService.cs

```diff
@@ -0,0 +1,408 @@
+                        cancellationToken,
+                        1000/*ms*/,
+                        TaskContinuationOptions.RunContinuationsAsynchronously,
+                        TaskScheduler.Default);
+                    _taskInFlight = true;
+                }
+            }
+
+            return Task.CompletedTask;
+        }
+
+        private async Task NotifyProjectSystemAsync(CancellationToken cancellationToken)
+        {
+            cancellationToken.ThrowIfCancellationRequested();
+
+            using var _1 = ArrayBuilder<DesignerInfo>.GetInstance(out var attributeInfos);
+            AddInfosAndResetQueue(attributeInfos);
+
+            // Reconcile all the notifications against the latest workspace solution we have.  This
+            // is technically racey as OOP may have computed against a different solution.  However,
+            // we should normally always reach a fixed point as we continually receive updates from
+            // OOP when the solution changes that we should normally be able to map to this
+            // solution.
+            var currentSolution = _workspace.CurrentSolution;
+
+            // Now, group all the notifications by project and update all the projects in parallel.
+            using var _2 = ArrayBuilder<Task>.GetInstance(out var tasks);
+            foreach (var group in attributeInfos.GroupBy(a => a.DocumentId.ProjectId))
+            {
+                cancellationToken.ThrowIfCancellationRequested();
+                tasks.Add(NotifyProjectSystemAsync(currentSolution, group.Key, group, cancellationToken));
+            }
+
+            // Wait until all project updates have happened before processing the next batch.
+            await Task.WhenAll(tasks).ConfigureAwait(false);
+        }
+
+        private void AddInfosAndResetQueue(ArrayBuilder<DesignerInfo> attributeInfos)
+        {
+            using var _ = PooledHashSet<DocumentId>.GetInstance(out var seenDocumentIds);
+
+            lock (_gate)
+            {
+                // walk the set of updates in reverse, and ignore documents if we see them a second
+                // time.  This ensures that if we're batching up multiple notifications for the same
+                // document, that we only bother processing the last one since it should beat out
+                // all the prior ones.
+                for (var i = _updatedInfos.Count - 1; i >= 0; i--)
+                {
+                    var designerArg = _updatedInfos[i];
+                    if (seenDocumentIds.Add(designerArg.DocumentId))
+                        attributeInfos.Add(designerArg);
+                }
+
+                // mark there being no existing update task so that the next OOP notification will
+                // kick one off.
+                _updatedInfos.Clear();
+                _taskInFlight = false;
+            }
+        }
+
+        private async Task NotifyProjectSystemAsync(
+            Solution solution,
```

Comment:
> Solution solution, [](start = 11, length = 19)

Can we eliminate access to the solution from this class? Seems like all we need is ProjectId and document file path.
We can include that information in the DesignerInfo.

Reply:
Checking!

Reply:
Worked.  Done!

Reply:
done.